test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/ValidationMessageTagHelperTest.cs

```diff
@@ -409,14 +409,76 @@ public class ValidationMessageTagHelperTest
 
             // Assert
             Assert.Equal("span", output.TagName);
-            Assert.Equal(2, output.Attributes.Count);
+            Assert.Equal(3, output.Attributes.Count);
             var attribute = Assert.Single(output.Attributes, attr => attr.Name.Equals("data-foo"));
             Assert.Equal("bar", attribute.Value);
             attribute = Assert.Single(output.Attributes, attr => attr.Name.Equals("data-hello"));
             Assert.Equal("world", attribute.Value);
+            attribute = Assert.Single(output.Attributes, attr => attr.Name.Equals("data-valmsg-replace"));
+            Assert.Equal("true", attribute.Value);
             Assert.Equal(expectedOutputContent, output.Content.GetContent());
         }
 
+        [Fact]
+        public void ProcessAsync_MergesTagBuilderFromGenerateWithoutValidationMessage()
+        {
+            // Arrange
+            var tagBuilder = new TagBuilder("span");
+            tagBuilder.InnerHtml.SetHtmlContent("");
+            tagBuilder.Attributes.Add("data-foo", "bar");
+            tagBuilder.Attributes.Add("data-hello", "world");
+            tagBuilder.Attributes.Add("data-valmsg-replace", "false");
+
+            var generator = new Mock<IHtmlGenerator>(MockBehavior.Strict);
+            var setup = generator
+                .Setup(mock => mock.GenerateValidationMessage(
+                    It.IsAny<ViewContext>(),
+                    It.IsAny<ModelExplorer>(),
+                    It.IsAny<string>(),
+                    It.IsAny<string>(),
+                    It.IsAny<string>(),
+                    It.IsAny<object>()))
+                .Returns(tagBuilder);
+
+            var validationMessageTagHelper = new ValidationMessageTagHelper(generator.Object)
+            {
+                For = CreateModelExpression("Hello")
+            };
+            var output = new TagHelperOutput(
+                "span",
+                attributes: new TagHelperAttributeList(),
+                getChildContentAsync: (useCachedResult, encoder) =>
+                {
+                    var tagHelperContent = new DefaultTagHelperContent();
+                    tagHelperContent.SetContent(null);
+                    return Task.FromResult<TagHelperContent>(tagHelperContent);
+                });
+
+            var context = new TagHelperContext(
+                tagName: "span",
+                allAttributes: new TagHelperAttributeList(
+                    Enumerable.Empty<TagHelperAttribute>()),
+                items: new Dictionary<object, object>(),
+                uniqueId: "test");
+
+            var viewContext = CreateViewContext();
+            validationMessageTagHelper.ViewContext = viewContext;
+
+            // Act
+            validationMessageTagHelper.ProcessAsync(context, output).Wait();
```

Comment:
If you're calling `ProcessAsync(...)`, use `await` and change the method's signature to include `async Task`